TestingLowerBounds/ForMathlib/LeftRightDeriv.lean

```diff
@@ -0,0 +1,190 @@
+      (Ne.symm (ne_of_lt hz)) hz'.le
+  exact MonotoneOn.tendsto_nhdsWithin_Ioi h_mono (bddBelow_slope_Ioi_of_convexOn hfc x)
+
+lemma differentiableWithinAt_Ioi_of_convexOn (x : ℝ) : DifferentiableWithinAt ℝ f (Ioi x) x :=
+  (hasRightDerivAt_of_convexOn hfc x).differentiableWithinAt
+
+lemma hadDerivWithinAt_rightDeriv_of_convexOn (x : ℝ) :
+    HasDerivWithinAt f (rightDeriv f x) (Ioi x) x :=
+  (differentiableWithinAt_Ioi_of_convexOn hfc x).hasDerivWithinAt
+
+lemma hasLeftDerivAt_of_convexOn (x : ℝ) :
+    HasDerivWithinAt f (sSup (slope f x '' Iio x)) (Iio x) x := by
+  simp_rw [hasDerivWithinAt_iff_tendsto_slope]
+  simp only [mem_Iio, lt_self_iff_false, not_false_eq_true, diff_singleton_eq_self]
+  have h_mono : MonotoneOn (slope f x) (Iio x) := by
+    refine monotoneOn_iff_forall_lt.mpr fun y hy z hz hz' ↦ ?_
+    simp_rw [slope_def_field]
+    exact ConvexOn.secant_mono hfc trivial trivial trivial (Ne.symm (ne_of_gt hy))
+      (Ne.symm (ne_of_gt hz)) hz'.le
+  exact MonotoneOn.tendsto_nhdsWithin_Iio h_mono (bddAbove_slope_Iio_of_convexOn hfc x)
+
+lemma differentiableWithinAt_Iio_of_convexOn (x : ℝ) : DifferentiableWithinAt ℝ f (Iio x) x :=
+  (hasLeftDerivAt_of_convexOn hfc x).differentiableWithinAt
+
+lemma hadDerivWithinAt_leftDeriv_of_convexOn (x : ℝ) :
+    HasDerivWithinAt f (leftDeriv f x) (Iio x) x :=
+  (differentiableWithinAt_Iio_of_convexOn hfc x).hasDerivWithinAt
+
+lemma rightDeriv_eq_sInf_slope_of_convexOn (x : ℝ) : rightDeriv f x = sInf (slope f x '' Ioi x) :=
+  (hasRightDerivAt_of_convexOn hfc x).derivWithin (uniqueDiffWithinAt_Ioi x)
+
+lemma leftDeriv_eq_sSup_slope_of_convexOn (x : ℝ) : leftDeriv f x = sSup (slope f x '' Iio x) :=
+  (hasLeftDerivAt_of_convexOn hfc x).derivWithin (uniqueDiffWithinAt_Iio x)
+
+lemma rightDeriv_mono : Monotone (rightDeriv f) := by
+  intro x y hxy
+  rcases eq_or_lt_of_le hxy with rfl | hxy; · rfl
+  simp_rw [rightDeriv_eq_sInf_slope_of_convexOn hfc]
+  refine csInf_le_of_le (b := slope f x y) (bddBelow_slope_Ioi_of_convexOn hfc x)
+    ⟨y, by simp [hxy]⟩ (le_csInf nonempty_of_nonempty_subtype ?_)
+  rintro _ ⟨z, yz, rfl⟩
+  rw [slope_comm]
+  exact slope_mono hfc trivial ⟨trivial, hxy.ne⟩ ⟨trivial, Ne.symm (ne_of_lt yz)⟩ (hxy.trans yz).le
+
+lemma leftDeriv_mono : Monotone (leftDeriv f) := by
+  rw [leftDeriv_eq_rightDeriv]
+  refine (Monotone.comp_antitone ?_ fun _ _ h ↦ neg_le_neg_iff.mpr h).neg
+  exact rightDeriv_mono (ConvexOn.comp_neg hfc)
+
+lemma leftDeriv_le_rightDeriv : leftDeriv f ≤ rightDeriv f := by
+  intro x
+  rw [rightDeriv_eq_sInf_slope_of_convexOn hfc, leftDeriv_eq_sSup_slope_of_convexOn hfc]
+  refine csSup_le nonempty_of_nonempty_subtype ?_
+  rintro _ ⟨z, (zx : z < x), rfl⟩
+  refine le_csInf nonempty_of_nonempty_subtype ?_
+  rintro _ ⟨y, xy, rfl⟩
+  exact slope_mono hfc trivial ⟨trivial, zx.ne⟩ ⟨trivial, (Ne.symm (ne_of_lt xy))⟩ (zx.trans xy).le
+
+lemma rightDeriv_right_continuous_of_convexOn (w : ℝ) :
```

Comment:
Is `right_continuous` a standard part of Mathlib lemma names about `ContinuousWithinAt` for `Ici`?

Reply:
It appears in Mathlib, but only [once](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Stieltjes.html#StieltjesFunction.right_continuous).